onnxruntime/test/providers/nnapi/nnapi_basic_test.cc

```diff
@@ -388,25 +433,117 @@ TEST(NnapiExecutionProviderTest, TestQDQReshape) {
 TEST(NnapiExecutionProviderTest, TestQDQSoftMax) {
   RunQDQModelTest(BuildQDQSoftMaxTestCase<uint8_t, uint8_t>(
                       {1, 32} /* input_shape */,
-                      static_cast<int64_t>(1) /* axis */),
+                      static_cast<int64_t>(1) /* axis */,
+                      1.f / 256 /* output_scales */,
+                      0 /* output_zp */),
                   "nnapi_qdq_test_graph_softmax",
                   {
                       true /* verify_entire_graph_use_ep */
                   });
 }
 
+// This is to verify when Nnapi required scale and zero point are not satisfied
+// the model can work as expected. (no nodes should be handled by Nnapi)
+TEST(NnapiExecutionProviderTest, TestQDQSoftMax_UnsupportedOutputScaleAndZp) {
+  RunQDQModelTest(BuildQDQSoftMaxTestCase<uint8_t, uint8_t>(
+                      {1, 32} /* input_shape */,
+                      static_cast<int64_t>(1) /* axis */,
+                      0.002f /* output_scales */,
+                      1 /* output_zp */),
+                  "nnapi_qdq_test_graph_softmax",
+                  {
+                      false /* verify_entire_graph_use_ep */,
+                      1e-5,
+                      true /* verify_expected_failure_graph */,
+                  },
+                  false /* nnapi_qdq_model_supported */);
+}
+
 TEST(NnapiExecutionProviderTest, TestQDQConcat) {
+  // This is to verify all the inputs have the same scale and zp as input 0
+  // See https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc#L16176
+  auto check_graph = [](Graph& test_graph) {
+    const auto* q_input_node = test_graph.GetNode(0);
+    const auto& input_def = q_input_node->InputDefs();
```

Comment:
>const auto& input_def = q_input_node->InputDefs(); [](http://example.com/codeflow?start=4&length=50)

this should be after the check that q_input_node != nullptr

these lines could potentially go into the helper function to extract scale/zp as well